include/clap/ext/draft/scratch-memory.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include "../../plugin.h"
+
+// This extension lets the plugin request "scratch" memory
+// from the host. Scratch memory can be accessed during the
+// `process()` callback, but is not persistent between callbacks.
+//
+// The motivation for this extension is to allow the plugin host
+// to "share" a single scratch buffer across multiple plugin
+// instances.
+//
+// For example, imagine the host needs to process three plugins
+// in sequence, and each plugin requires 10K of scratch memory.
+// If each plugin pre-allocates its own scratch memory, then 30K
+// of memory is being allocated in total. However, if each plugin
+// requests 10K of scratch memory from the host, then the host can
+// allocate a single 10K scratch buffer, and make it available to all
+// three plugins.
+//
+// On memory-constrained platforms, this optimization may allow for
+// more plugins to be used simultaneously. On platforms with lots
+// of memory, this optimization may improve CPU cache usage.
+
+static CLAP_CONSTEXPR const char CLAP_EXT_SCRATCH_MEMORY[] = "clap.scratch-memory/1";
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct clap_host_scratch_memory {
+   // Asks the host for certain amount of scratch memory.
+   // If the host is unable to provide the memory, it should
+   // return "false".
+   //
+   // Note that any memory the host allocates to satisfy
+   // the requested scratch size can be de-allocated by the
+   // host when the plugin is de-activated.
+   //
+   // In the context of plugins and hosts that implement
+   // the "thread-pool" extension, scratch memory is assumed
+   // to be "thread-local". The plugin should request the maximum
+   // amount of scratch memory that it will need on a single
+   // thread. Accordingly, the host must ensure that each
+   // thread can independently provide the requested amount
+   // of scratch memory.
+   //
+   // [main-thread & being-activated]
+   bool(CLAP_ABI *reserve)(const clap_host_t *host, uint32_t scratch_size_bytes);
+
+   // Asks the host for the previously requested scratch memory.
+   // If the host returned "true" when scratch memory was requested,
+   // then this method must return a pointer to a memory block at least
+   // as large as the requested size. If the host returned "false"
+   // when scratch memory was requested then this method should return
+   // null.
```

Comment:
```suggestion
   // Asks the host for the previously reserved scratch memory.
   // If the host returned "true" when scratch memory was requested,
   // then this method must return a pointer to a memory block at least
   // as large as the reserved size. If the host returned "false"
   // when scratch memory was requested, then this method must not
   // be called.
```
Just somehow it could be made more clear that it's related to the `reserve` function.

Also I think calling it when the host already said "no" makes no sense and just should be forbidden (?)

Reply:
Thanks for the edits :). `must not be called` is maybe a little strict, but I think I'm in favor of it.